scrape/scrape.go

```diff
@@ -630,6 +634,14 @@ func appender(app storage.Appender, sampleLimit, bucketLimit int) storage.Append
 			limit:    bucketLimit,
 		}
 	}
+
+	if maxSchema < nativeHistogramMaxSchema {
+		app = &maxSchemaAppender{
+			Appender:  app,
+			maxSchema: maxSchema,
+		}
+	}
```

Comment:
I'm a bit worried that conservative users might often use both, a bucket limit _and_ a schema limit. In this case, we might need to call `ReduceResolution` twice. How about combining both `bucketLimitAppender` and `maxSchemaAppender` into one `limitResolutionAppender`, which calculates the required schema based on both config settings and then runs `ReduceResolution` at most once?

Reply:
> I'm a bit worried that conservative users might often use both, a bucket limit _and_ a schema limit. In this case, we might need to call `ReduceResolution` twice. How about combining both `bucketLimitAppender` and `maxSchemaAppender` into one `limitResolutionAppender`, which calculates the required schema based on both config settings and then runs `ReduceResolution` at most once?

I think it would be impossible to merge `bucketLimitAppender` and `maxSchemaAppender` into a single `limitResolutionAppender` while ensuring that `ReduceResolution` runs at most once. 

This is because we seek a maximum schema value that simultaneously satisfies both the `maxSchemaAppender` and the `bucketLimitAppender`. The most intuitive way to obtain this is by:
- First, calculating the maximum schema value that satisfies `maxSchemaAppender`, let's call it `ms`.
- Then, calculating the maximum schema value that satisfies `bucketLimitAppender`, let's call it `mb`. 
- Afterward, we take the minimum of `ms` and `mb`.

Since the method that `bucketLimitAppender` uses to calculate the maximum schema value depends on iteratively calling the `ReduceResolution` function, decrementing the schema by one each time until it finds the schema value `mb` that meets the requirements. In contrast, `maxSchemaAppender` computes the maximum schema in a more efficient manner, by calling `ReduceResolution` just once if the current schema is greater than `maxSchema`.

Therefore, to minimize calls to `ReduceResolution`, we must make an effort to first calculate the schema `ms` using `maxSchemaAppender`, and then check whether this schema meets the requirements of `bucketLimitAppender`. If it doesn't, we still need to rely on `bucketLimitAppender` to repeatedly call `ReduceResolution` to determine the correct result.

@beorn7 WDYT?

Reply:
Sounds very reasonable. Thanks for the analysis.